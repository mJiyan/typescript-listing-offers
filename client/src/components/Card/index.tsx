import React from 'react';
import { OfferType } from '../../redux/actions/Offer/types';
import fallBackImage from '../../assets/images/fallbackImage.png';

const Card: React.FC<OfferType> = ({ headlines, prices, splashImages }) => {
  return (
    <div className="flex items-center justify-center p-3">
      <div className="rounded overflow-hidden shadow-lg h-64 w-64 bg-gray-100">
        {splashImages !== undefined ? (
          <img className="w-full" src={splashImages?.[0].wideMedium} />
        ) : (
          <img className="w-full" src={fallBackImage} />
        )}
        <div className="flex flex-col items-center justify-center py-12">
          <p className="text-orange-600 text-lg "> {headlines.description}</p>
          <p className="text-sm">
            {prices.totalPrice.amount.display} {prices.totalPrice.amount.currency}
          </p>
        </div>
      </div>
    </div>
  );
};

export default Card;
