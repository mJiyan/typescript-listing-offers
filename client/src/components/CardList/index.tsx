import React from 'react';
import { OfferType } from '../../redux/actions/Offer/types';
import { Card } from '..';
import { OFFER_PER_PAGE } from '../../services/constants';

type Props = {
  offers: Array<OfferType>;
  page: number;
};

const CardList: React.FC<Props> = ({ offers, page }) => {

  // this function is returning a selected page number
  const calculateIndex = (page: number): number => {
    return (page - 1) * OFFER_PER_PAGE;
  };

  // this function is returning a selected pages list
  const paginateOffers = (startIndex: number): Array<OfferType> => {
    return offers.slice(startIndex, startIndex + OFFER_PER_PAGE);
  };

  const startIndex = calculateIndex(page);
  const selectedOffers = paginateOffers(startIndex);

  return (
    <div className="grid xl:grid-cols-6 lg:grid-cols-3 md:grid-cols-2 sm:grid-cols-1 xs:grid-cols-1 items-center xl:mb-auto xl:h-10 ">
      {selectedOffers.map((offer: OfferType, index: number) => {
        return (
          <Card
            key={index}
            headlines={offer.headlines}
            prices={offer.prices}
            splashImages={offer.splashImages}
          />
        );
      })}
    </div>
  );
};

export default CardList;
