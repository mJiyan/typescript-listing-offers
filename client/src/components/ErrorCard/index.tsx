import Exclamation from '../../assets/images/exclamation.png';

const ErrorCard = () => {
  return (
    <div className="flex items-center justify-center h-screen">
      <section className="text-center">
        <div className="max-w-xs mx-auto px-16">
          <img src={Exclamation} />
        </div>
        <h2 className="mt-8 uppercase text-xl font-black">
          We're sorry, but something went wrong.
        </h2>
        <p className="mt-6 text-sm text-gray-900">
          If you are the application owner check the logs for more information.
        </p>
      </section>
    </div>
  );
};

export default ErrorCard;
