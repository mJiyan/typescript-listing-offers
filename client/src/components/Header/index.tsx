import React from 'react';
import Logo from '../../assets/images/sixt-logo.png';

const Header = () => {
  return (
    <nav className="bg-orange-600 h-24">
      <div className=" flex items-center justify-center flex-wrap py-2">
        <span className="font-semibold text-sm tracking-tight flex flex-col items-center">
          <img src={Logo} className="h-16 w-16" />
          <p className='px-5'> Rent a car</p>
        </span>
      </div>
    </nav>
  );
};

export default Header;
