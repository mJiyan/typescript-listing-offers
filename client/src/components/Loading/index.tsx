import React from 'react';

const Loading = () => {
  return (
    <div className="flex items-center justify-center align-center grid xl:grid-cols-6 lg:grid-cols-3 md:grid-cols-2 sm:grid-cols-2 xs:grid-cols-1 xl:mb-auto xl:h-10 lg:mb-auto lg:h-10">
      <div className="p-3 h-64 w-64  animate-pulse">
        <div className="rounded overflow-hidden shadow-lg ">
          <div className="bg-gray-100 h-32 w-full"></div>

          <div className="px-6 py-6 ">
            <div className="font-bold text-sm mb-2 bg-gray-100">&nbsp;</div>
            <p className="text-gray-100 text-xs bg-gray-100">&nbsp;</p>
          </div>
        </div>
      </div>
      <div className="p-3 h-64 w-64  animate-pulse">
        <div className="rounded overflow-hidden shadow-lg ">
          <div className="bg-gray-100 h-32 w-full"></div>

          <div className="px-6 py-6 ">
            <div className="font-bold text-sm mb-2 bg-gray-100">&nbsp;</div>
            <p className="text-gray-100 text-xs bg-gray-100">&nbsp;</p>
          </div>
        </div>
      </div>

      <div className="p-3 h-64 w-64  animate-pulse">
        <div className="rounded overflow-hidden shadow-lg ">
          <div className="bg-gray-100 h-32 w-full"></div>

          <div className="px-6 py-6 ">
            <div className="font-bold text-sm mb-2 bg-gray-100">&nbsp;</div>
            <p className="text-gray-100 text-xs bg-gray-100">&nbsp;</p>
          </div>
        </div>
      </div>

      <div className="p-3 h-64 w-64  animate-pulse">
        <div className="rounded overflow-hidden shadow-lg ">
          <div className="bg-gray-100 h-32 w-full"></div>

          <div className="px-6 py-6 ">
            <div className="font-bold text-sm mb-2 bg-gray-100">&nbsp;</div>
            <p className="text-gray-100 text-xs bg-gray-100">&nbsp;</p>
          </div>
        </div>
      </div>

      <div className="p-3 h-64 w-64  animate-pulse">
        <div className="rounded overflow-hidden shadow-lg ">
          <div className="bg-gray-100 h-32 w-full"></div>

          <div className="px-6 py-6 ">
            <div className="font-bold text-sm mb-2 bg-gray-100">&nbsp;</div>
            <p className="text-gray-100 text-xs bg-gray-100">&nbsp;</p>
          </div>
        </div>
      </div>

      <div className="p-3 h-64 w-64  animate-pulse">
        <div className="rounded overflow-hidden shadow-lg ">
          <div className="bg-gray-100 h-32 w-full"></div>

          <div className="px-6 py-6 ">
            <div className="font-bold text-sm mb-2 bg-gray-100">&nbsp;</div>
            <p className="text-gray-100 text-xs bg-gray-100">&nbsp;</p>
          </div>
        </div>
      </div>

    </div>
  );
};

export default Loading;
