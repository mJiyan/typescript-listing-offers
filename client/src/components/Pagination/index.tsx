import React from 'react';

type Props = {
  totalPages: number;
  changePage: (pageNumber: number) => void;
  page: number;
};

// with the changePage function (which is passed from Offers) changing the current page for the list
const Pagination: React.FC<Props> = ({ totalPages, changePage, page }) => {
  // this function creating number array for the pagination component's button
  const convertTotalPagesToArray = (totalPage: number): Array<number> => {
    return [...Array(totalPage).keys()].map((num) => num + 1);
  };
  const pages = convertTotalPagesToArray(totalPages);

  return (
    <div className="flex flex-wrap justify-center  xl:h-10  xl:mb-20  ">
      <ul className="inline-flex" data-testid="list-page-buttons">
        <li>
          <button
            data-testid="prev-button"
            disabled={page === 1 && true}
            className="h-10 px-5 text-orange-600 transition-colors duration-150 bg-white rounded-l-lg focus:shadow-outline hover:bg-orange-100"
            onClick={() => changePage(page - 1)}
          >
            Prev
          </button>
        </li>
        {pages.map((num) => (
          <li key={num}>
            <button
              key={num}
              data-testid={`${num}-button`}
              onClick={() => changePage(num)}
              className={`h-10 px-2 duration-150  focus:shadow-outline hover:bg-orange-100  rounded-full ${
                num === page ? 'bg-orange-600 text-white-100' : 'text-orange-600'
              }`}
            >
              {num}
            </button>
          </li>
        ))}
        <li>
          <button
            data-testid="next-button"
            className="h-10 px-5 text-orange-600 transition-colors duration-150 bg-white rounded-r-lg focus:shadow-outline hover:bg-orange-100"
            onClick={() => changePage(page + 1)}
            disabled={page === pages.length && true}
          >
            Next
          </button>
        </li>
      </ul>
    </div>
  );
};

export default Pagination;
