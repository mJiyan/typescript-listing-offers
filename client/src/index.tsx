import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import store from './Store';
import { Offers } from './views';
import './assets/style/tailwind.css';

ReactDOM.render(
  <Provider store={store}>
    <Offers />
  </Provider>,
  document.getElementById('root'),
);
