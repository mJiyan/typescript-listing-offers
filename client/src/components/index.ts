import Card from './Card';
import ErrorCard from './ErrorCard';
import Loading from './Loading';
import Header from './Header';
import CardList from './CardList';
import Pagination from './Pagination';

export { Card, ErrorCard, Loading, Header, CardList, Pagination };
