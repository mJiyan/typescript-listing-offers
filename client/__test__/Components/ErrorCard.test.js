import React from 'react';
import { ErrorCard } from '../../src/components';
import { mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import defaultState from '../Mock/defaulState';
import 'jsdom-global/register'; //at the top of file , even  , before importing react
import toJson from 'enzyme-to-json'; //added this line

const mockStore = configureMockStore([thunk]);
const store = mockStore(defaultState);

const wrapper = mount(
  <Provider store={store}>
    <ErrorCard />
  </Provider>,
);

describe('rendering components', () => {
  test('it should match with the snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('it should contain with the header', () => {
    const header = (
      <h2 className="mt-8 uppercase text-xl font-black">We're sorry, but something went wrong.</h2>
    );
    expect(wrapper.contains(header)).toEqual(true);
  });
});
