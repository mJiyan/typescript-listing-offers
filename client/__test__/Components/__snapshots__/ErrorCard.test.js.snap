// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering components it should match with the snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ErrorCard>
    <div
      className="flex items-center justify-center h-screen"
    >
      <section
        className="text-center"
      >
        <div
          className="max-w-xs mx-auto px-16"
        >
          <img
            src={
              Object {
                "process": [Function],
              }
            }
          />
        </div>
        <h2
          className="mt-8 uppercase text-xl font-black"
        >
          We're sorry, but something went wrong.
        </h2>
        <p
          className="mt-6 text-sm text-gray-900"
        >
          If you are the application owner check the logs for more information.
        </p>
      </section>
    </div>
  </ErrorCard>
</Provider>
`;
